### PR TITLE
Move feature transformer to pyterrier_util

### PR DESCRIPTION
**Pull request by gijshendriksen**
This PR reduces code duplication between `tira.pt.doc_features()` and `tira.pt.query_features()`, and makes it easier to add new query-only, doc-only or query-document features.

The new `TiraApplyFeatureTransformer` accepts a `mapping` parameter that assigns input rows to a feature or feature vector. For instance, if we have a mapping with `mapping['q0']['doc0'] = np.array([.1, .2])`, we can instantiate the transformer as `TiraApplyFeatureTransformer(mapping, ('qid', 'docno'))`. For one row, it will then create the features by selecting `mapping[row['qid']][row['docno']]`.

@mam10eks if that's alright with you, I will expand the `doc_features` and `query_features` unit tests later, after the LongEval deadline. Do I also need to update or add documentation somewhere?

### CODE REVIEW

File: python-client/tira/pyterrier_util.py [hunk 1]
```diff
@@ -148,3 +148,24 @@ def transform(self, topics):
...
+        outputRes = inputRes.copy()
+        outputRes["features"] = outputRes.apply(self._get_features_from_row, axis=1)
+        return outputRes
+
+    def _get_features_from_row(self, row):
+        value = self.mapping
+        for col in self.cols:
+            value = value[str(row[col])]
+
+        return value
```
Comment (TheMrSheldon) [inline python-client/tira/pyterrier_util.py:171]: What value will `value` have at the end of the function if `cols` held multiple values?

**Comment by TheMrSheldon** [comment]
Thank you very much for the PR! Could you add some short documentation and type hinting to `TiraApplyFeatureTransformer`? I don't find it immediately obvious what it does or why it is equivalent to the code it replaces.

**Comment by gijshendriksen** [comment]
Hi @TheMrSheldon, thank you for the quick feedback! I have added some documentation to the `TiraApplyFeatureTransformer` class, which hopefully explains what it's supposed to achieve. If this is still unclear, or the `_get_features_from_row` method requires additional docstrings, please let me know!

File: python-client/tira/pyterrier_util.py (hunk 1 above)
Comment (gijshendriksen) [reply python-client/tira/pyterrier_util.py]: The method will iterate through levels of the mapping, until it ends up with the final feature value. Assuming that `self.mapping` is as deep as `self.cols` is long, the final value of `value` will thus be a 'leaf' of the mapping, i.e. the final feature vector.

If `mapping = {'q0': {'d0': np.array([0.1, 0.2])}}`, and we'd have a dataframe with e.g. rows `qid,query,docno,text`, we'd set `cols=('qid','docno')` and we'd iteratively compute:
```
value_0 = {'q0': {'d0': np.array([0.1, 0.2])}}
value_1 = value_0['q0']
value_1 = {'d0': np.array([0.1, 0.2])}
value_2 = value_1['d0']
value_2 = np.array([0.1, 0.2]
```
So the return value would be a 2D feature vector. Hope that helps clear things up!

Comment (TheMrSheldon) [reply python-client/tira/pyterrier_util.py]: Thank you very much for the clarification and the documentation!

**Review by TheMrSheldon — APPROVED** [review]
Approved, you can go ahead and merge :)

**Comment by gijshendriksen** [comment]
Thank you for the quick review!

**Pull request merged**